### ISSUE TITLE
[Feature]: Perceptron

### ISSUE THREAD

**Issue by kokou-egbewatt** [kokou-egbewatt/deep-learning#2]
### 📋 Issue Type

Select the type of issue (check one):

- [ ] 🐛 Bug Report - Something isn't working correctly
- [x] ✨ Feature Request - New neural network architecture or algorithm
- [ ] 📚 Documentation - Improvements to notebooks or documentation
- [ ] ❓ Question - Help understanding an implementation
- [ ] 🔧 Enhancement - Improve existing implementation

______________________________________________________________________

### 📝 Description

Implement a single-layer perceptron from scratch for binary classification. The implementation should include the classic perceptron learning algorithm with comprehensive visualizations showing how the network learns over time.

______________________________________________________________________

### 🎯 Context (for Feature Requests)

**Algorithm/Architecture:** Single-layer perceptron with hard-limit activation function

**Why it's valuable:**
- Foundation of neural networks - essential for understanding more complex architectures
- Demonstrates core concepts: weights, bias, activation functions, learning rules
- Shows geometric intuition through decision boundary visualization
- Historically significant algorithm with provable convergence properties

**Use cases:**
- Binary classification on linearly separable data
- Educational demonstration of gradient-free learning
- Visualization of training dynamics and convergence
- Foundation for understanding multi-layer perceptrons

______________________________________________________________________

### 🐛 Bug Details (if applicable)

N/A - This is a feature request

______________________________________________________________________

### 📊 Neural Network Components

Which components are affected or relevant? (check all that apply)

- [x] Perceptron / Single-layer networks
- [ ] Multi-layer perceptrons (MLP)
- [x] Activation functions (hard-limit/step function)
- [ ] Loss functions
- [x] Training algorithms (perceptron learning rule)
- [x] Visualization tools (weight evolution, decision boundaries)
- [x] Data generation/preprocessing (linearly separable 2D data)
- [x] Notebooks (Jupyter notebook with explanations)
- [x] Documentation (markdown explanations, docstrings)

______________________________________________________________________

### 🔬 Environment

- Python version: 3.12+
- NumPy version: 1.24.0+
- Matplotlib version: 3.7.0+
- Operating System: Cross-platform (Windows, macOS, Linux)

______________________________________________________________________

### 💡 Suggested Implementation (Optional)

```python
class SimplePerceptron:
    """Single-layer perceptron for binary classification."""
    
    def __init__(self, n_features, learning_rate=1.0):
        self.weights = np.zeros(n_features)
        self.bias = 0.0
        self.learning_rate = learning_rate
        # Track history for visualization
        self.weight_history = []
        self.bias_history = []
        self.epoch_errors = []
        self.epoch_accuracy = []
    
    def hard_limit(self, net_input):
        """Step activation: 1 if net_input >= 0, else 0"""
        return 1 if net_input >= 0 else 0
    
    def predict(self, X):
        """Make predictions on input data."""
        net_input = np.dot(X, self.weights) + self.bias
        return np.array([self.hard_limit(x) for x in net_input])
    
    def train(self, X, y, epochs=100):
        """Train using perceptron learning rule: Δw = (t - a) * x"""
        for epoch in range(epochs):
            epoch_error = 0
            for x_i, y_i in zip(X, y):
                prediction = self.predict(x_i.reshape(1, -1))[0]
                error = y_i - prediction
                epoch_error += int(error != 0)
                
                # Update: w_new = w_old + lr * error * input
                self.weights += self.learning_rate * error * x_i
                self.bias += self.learning_rate * error
            
            # Track metrics
            self.epoch_errors.append(epoch_error)
            self.weight_history.append(self.weights.copy())
            
            # Early stopping
            if epoch_error == 0:
                break
        
        return self
```

**References:**
- [x] [MathWorks Perceptron Documentation](https://www.mathworks.com/help/deeplearning/ug/perceptron-neural-networks.html)
- [x] Rosenblatt, F. (1958). The perceptron: A probabilistic model for information storage
- [x] Modern NumPy random generation patterns
